jwxml/siaf.py

```diff
@@ -0,0 +1,739 @@
+            ap.plotDetectorOrigin(frame=frame, which=which)
+
+    def plotDetectorChannels(self, frame=None):
+        """ Mark on the plot the various detector readout channels
+
+        These are depicted as alternating light/dark bars to show the
+        regions read out by each of the output amps.
+
+        Parameters
+        ----------
+        frame : str
+            Which coordinate system to plot in: 'Tel', 'Idl', 'Sci', 'Det'
+            Optional if you have already called plot() to specify a
+            coordinate frame.
+
+        """
+        if frame is None: frame = self._last_plot_frame
+        for ap in self._getFullApertures():
+            ap.plotDetectorChannels(frame=frame)
+
+
+def plotAllSIAFs(subarrays=True, showorigin=True, showchannels=True, **kwargs):
+    """ Plot All instrument """
+
+    for instr in ['NIRCam', 'NIRISS', 'NIRSpec', 'FGS', 'MIRI']:
+        aps = SIAF(instr, **kwargs)
+        print("{0} has {1} apertures".format(aps.instrument, len(aps)))
+
+        aps.plot(clear=False, subarrays=subarrays, **kwargs)
+        if showorigin: aps.plotDetectorOrigin()
+        if showchannels: aps.plotDetectorChannels()
+
+
+def plotMainSIAFs(label=False, darkbg=False, **kwargs):
+    if darkbg:
+        col_imaging = 'aqua'
+        col_coron = 'lime'
+        col_msa = 'violet'
+    else:
+        col_imaging = 'blue'
+        col_coron = 'green'
+        col_msa = 'magenta'
+
+    nircam = SIAF('NIRCam')
+    niriss = SIAF('NIRISS')
+    fgs = SIAF('FGS')
+    nirspec = SIAF('NIRSpec')
+    miri = SIAF('MIRI')
+
+    im_aps = [
+        nircam['NRCA5_FULL'],
+        nircam['NRCB5_FULL'],
+        niriss['NIS-CEN'],
+        miri['MIRIM_ILLUM'],
+        fgs['FGS1_FULL'],
+        fgs['FGS2_FULL']
+    ]
+
+    for letter in ['A', 'B']:
+        for num in range(5):
+            im_aps.append(nircam['NRC{}{}_FULL'.format(letter, num + 1)])
+
+    coron_aps = [
+        nircam['NRCA2_MASK210R'],
+        nircam['NRCA4_MASKSWB'],
+        nircam['NRCA5_MASK335R'],
+        nircam['NRCA5_MASK430R'],
+        nircam['NRCA5_MASKLWB'],
+        nircam['NRCB3_MASKSWB'],
+        nircam['NRCB1_MASK210R'],
+        nircam['NRCB5_MASK335R'],
+        nircam['NRCB5_MASK430R'],
+        nircam['NRCB5_MASKLWB'],
+        miri['MIRIM_MASK1065'],
+        miri['MIRIM_MASK1140'],
+        miri['MIRIM_MASK1550'],
+        miri['MIRIM_MASKLYOT']
+    ]
+    msa_aps = [nirspec['NRS_FULL_MSA' + str(n + 1)] for n in range(4)]
+    msa_aps.append(nirspec['NRS_S1600A1_SLIT'])  # square aperture
+
+    for aplist, col in zip([im_aps, coron_aps, msa_aps], [col_imaging, col_coron, col_msa]):
+        for ap in aplist:
+            ap.plot(color=col, frame='Tel', label=label, **kwargs)
+
+    # ensure V2 increases to the left
+    ax = plt.gca()
+    xlim = ax.get_xlim()
+    if xlim[0] < xlim[1]:
+        ax.set_xlim(xlim[::-1])
+
+
+class Test_SIAF(unittest.TestCase):
+    def assertAlmostEqualTwo(self, tuple1, tuple2):
```

Comment:
Should we relocate this to `tests/test_siaf.py`? It will also need updating to fix the current hard coded filename for a six year old SIAF instance. But that can be a separate PR after this one of course.